presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -79,31 +91,163 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+                return Optional.empty();
+            }
+
+            DecorrelationResult childDecorrelationResult = childDecorrelationResultOptional.get();
+            if (childDecorrelationResult.atMostSingleRow) {
+                return childDecorrelationResultOptional;
+            }
+
+            if (node.getCount() != 1) {
+                return Optional.empty();
+            }
+
+            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+            if (constantSymbols.isEmpty() || !constantSymbols.containsAll(node.getOutputSymbols())) {
+                return Optional.empty();
+            }
+
+            // Rewrite to aggregation on symbols which we know are constants and are also part of output
+            AggregationNode aggregationNode = new AggregationNode(
+                    idAllocator.getNextId(),
+                    node.getSource(),
+                    ImmutableMap.of(),
+                    ImmutableList.of(ImmutableList.<Symbol>builder()
+                            .addAll(constantSymbols)
+                            .addAll(node.getOutputSymbols())
+                            .build()),
+                    AggregationNode.Step.SINGLE,
+                    Optional.empty(),
+                    Optional.empty());
+
+            childDecorrelationResultOptional = aggregationNode.accept(this, null);
+            if (!childDecorrelationResultOptional.isPresent() || !childDecorrelationResultOptional.get().atMostSingleRow) {
+                return Optional.empty();
+            }
+
+            return childDecorrelationResultOptional;
+        }
+
+        @Override
+        public Optional<DecorrelationResult> visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
+        {
+            Optional<DecorrelationResult> childDecorrelationResultOptional = lookup.resolve(node.getSource()).accept(this, null);
+            if (!childDecorrelationResultOptional.isPresent() || !childDecorrelationResultOptional.get().atMostSingleRow) {
+                return Optional.empty();
+            }
+
+            return childDecorrelationResultOptional;
+        }
+
+        @Override
+        public Optional<DecorrelationResult> visitAggregation(AggregationNode node, Void context)
+        {
+            if (node.hasEmptyGroupingSet()) {
+                return Optional.empty();
+            }
+
+            Optional<DecorrelationResult> childDecorrelationResultOptional = lookup.resolve(node.getSource()).accept(this, null);
+            if (!childDecorrelationResultOptional.isPresent()) {
+                return Optional.empty();
+            }
+
+            DecorrelationResult childDecorrelationResult = childDecorrelationResultOptional.get();
+            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+
+            AggregationNode decorrelatedAggregation = childDecorrelationResult.getCorrelatedSymbolMapper()
```

Comment:
Maybe instead of using mapper, you could inline its `map` method here and update grouping sets in one step. It is a pity that you need to create intermediate aggregation node which is incorrect due missing symbols in grouping sets.

Reply:
> It is a pity that you need to create intermediate aggregation node which is incorrect due missing symbols in grouping sets.

It's not incorrect. It just contains `correlated symbols`, so we need to decorrelate it.